features/phpunit_exceptions.feature

```diff
@@ -0,0 +1,244 @@
+Feature: Stringifying PHPUnit exceptions
+  In order to understand why a step has failed
+  As a feature developer
+  I need to see the details of failed PHPUnit assertions if I am using a supported version
+
+  Background:
+      Given a file named "features/bootstrap/FeatureContext.php" with:
+      """
+      <?php
+
+      use Behat\Behat\Context\Context;
+      use Behat\Hook\BeforeFeature;
+      use Behat\Hook\BeforeSuite;
+      use Behat\Step\Then;
+      use PHPUnit\Framework\Assert;
+
+      class FeatureContext implements Context
+      {
+
+          #[BeforeSuite]
+          public static function beforeSuiteEnableNativeAssert(): void
+          {
+              // Enforce native assertions during this run, so we can use them to check state without using PHPUnit classes.
+              ini_set('assert.active', true);
+              ini_set('assert.exception', true);
+          }
+
+          #[BeforeFeature('@phpunit_10_broken')]
+          public static function beforeFeatureBreakPHPUnit10(): void
+          {
+              // Note this test proves both that we're handling exceptions, and that Behat will use the PHPUnit 10
+              // ThrowableToStringMapper class if it's present - even though at the moment we're installing PHPUnit 9.
+              static::assertClassNotLoaded(\PHPUnit\Util\ThrowableToStringMapper::class);
+              require_once(__DIR__.'/IncompatibleThrowableToStringMapper.php');
+              class_alias(IncompatibleThrowableToStringMapper::class, \PHPUnit\Util\ThrowableToStringMapper::class);
+          }
+
+          #[BeforeFeature('@phpunit_incompatible')]
+          public static function beforeFeatureRemoveKnownPHPUnit(): void
+          {
+              // At the start of the feature, this Behat process should not have referenced any PHPUnit classes.
+              // So the easiest way to simulate an incompatible PHPUnit version is to wrap the registered autoloader(s)
+              // and prevent PHP from finding the exception formatting classes we support.
+              // This will only affect the Behat process we're testing - the outer test runner will find them as usual.
+              static::assertClassNotLoaded(\PHPUnit\Util\ThrowableToStringMapper::class);
+              static::assertClassNotLoaded(\PHPUnit\Framework\TestFailure::class);
+
+              if (PHP_VERSION_ID < 80400) {
+                  // Trigger loading array_find from symfony/polyfill-php84 before our autoloader tries to use it
```

Comment:
there is no function autoloading in PHP, so this is totally useless. The function polyfills are loaded eagerly at the time you load the `vendor/autoload.php` file

Reply:
@stof the functions are autoloaded but until you call them they don't try to load the Php84 class that the polyfill actually uses. I know because without this the autoloader failed with an error due to not autoloading the symfony class.

Reply:
@stof I was on the move earlier, [here](https://github.com/symfony/polyfill/blob/a5ae13211e892f1b1e955f9a22d29eb1f9581c1f/src/Php84/bootstrap.php#L26-L28
) is how symfony/polyfill implements array_find:

```
if (!function_exists('array_find')) {
    function array_find(array $array, callable $callback) { return p\Php84::array_find($array, $callback); }
}
```

As you can see, if this is the first Php84 polyfill that has been called, it will trigger an autoload of the Php84 class. Not normally an issue, but if you're using array_find *inside* an autoloader that can be a problem as indeed it did here.

FWIW, it may be just a language thing, or that I am tired after a long week, but I read your comment as quite rude. From my perspective it didn't feel great to have you engage with a PR that has involved a fair bit of effort and thought, and that fixes a longstanding issue, and your only comment is to describe a tiny part of it as "totally useless".

I flag this less for my own sake, more that if we want to get behat moving again I hope we can all try to make contributors feel welcome and valued.